sdks/go/pkg/beam/registration/emitterIterRegistration.go

```diff
@@ -0,0 +1,298 @@
+	return e.fn
+}
+
+func (e *emitNative1[T]) invoke(val T) {
+	e.value = exec.FullValue{Windows: e.ws, Timestamp: e.et, Elm: val}
+	if err := e.n.ProcessElement(e.ctx, &e.value); err != nil {
+		panic(err)
+	}
+}
+
+type emitNative2[T1, T2 any] struct {
+	n  exec.ElementProcessor
+	fn interface{}
+
+	ctx   context.Context
+	ws    []typex.Window
+	et    typex.EventTime
+	value exec.FullValue
+}
+
+func (e *emitNative2[T1, T2]) Init(ctx context.Context, ws []typex.Window, et typex.EventTime) error {
+	e.ctx = ctx
+	e.ws = ws
+	e.et = et
+	return nil
+}
+
+func (e *emitNative2[T1, T2]) Value() interface{} {
+	return e.fn
+}
+
+func (e *emitNative2[T1, T2]) invoke(key T1, val T2) {
+	e.value = exec.FullValue{Windows: e.ws, Timestamp: e.et, Elm: key, Elm2: val}
+	if err := e.n.ProcessElement(e.ctx, &e.value); err != nil {
+		panic(err)
+	}
+}
+
+type emitNative2WithTimestamp[T any] struct {
+	n  exec.ElementProcessor
+	fn interface{}
+
+	ctx   context.Context
+	ws    []typex.Window
+	et    typex.EventTime
+	value exec.FullValue
+}
+
+func (e *emitNative2WithTimestamp[T]) Init(ctx context.Context, ws []typex.Window, et typex.EventTime) error {
+	e.ctx = ctx
+	e.ws = ws
+	e.et = et
+	return nil
+}
+
+func (e *emitNative2WithTimestamp[T]) Value() interface{} {
+	return e.fn
+}
+
+func (e *emitNative2WithTimestamp[T]) invoke(et typex.EventTime, val T) {
+	e.value = exec.FullValue{Windows: e.ws, Timestamp: et, Elm: val}
+	if err := e.n.ProcessElement(e.ctx, &e.value); err != nil {
+		panic(err)
+	}
+}
+
+type emitNative3[T1, T2 any] struct {
+	n  exec.ElementProcessor
+	fn interface{}
+
+	ctx   context.Context
+	ws    []typex.Window
+	et    typex.EventTime
+	value exec.FullValue
+}
+
+func (e *emitNative3[T1, T2]) Init(ctx context.Context, ws []typex.Window, et typex.EventTime) error {
+	e.ctx = ctx
+	e.ws = ws
+	e.et = et
+	return nil
+}
+
+func (e *emitNative3[T1, T2]) Value() interface{} {
+	return e.fn
+}
+
+func (e *emitNative3[T1, T2]) invoke(et typex.EventTime, key T1, val T2) {
+	e.value = exec.FullValue{Windows: e.ws, Timestamp: et, Elm: key, Elm2: val}
+	if err := e.n.ProcessElement(e.ctx, &e.value); err != nil {
+		panic(err)
+	}
+}
+
+// RegisterEmitter1 registers parameters from your DoFn with a
+// signature func(T) and optimizes their execution.
+// This must be done by passing in a reference to an instantiated version
+// of the function, aka:
+// registration.RegisterEmitter1[T]((*func(T))(nil))
+func RegisterEmitter1[T1 any](e *func(T1)) {
+	registerFunc := func(n exec.ElementProcessor) exec.ReusableEmitter {
+		gen := &emitNative1[T1]{n: n}
+		gen.fn = gen.invoke
+		return gen
+	}
+	exec.RegisterEmitter(reflect.TypeOf(e).Elem(), registerFunc)
+}
+
+// RegisterEmitter2 registers parameters from your DoFn with a
+// signature func(T1, T2) and optimizes their execution.
+// This must be done by passing in a reference to an instantiated version
+// of the function, aka:
+// registration.RegisterEmitter2[T1, T2]((*func(T1, T2))(nil))
```

Comment:
It may be worth reinforcing the assumed order of return values in functions here and in RegisterEmitter3 in some way since it's another place we want EventTime values first.

Reply:
I think function registration takes care of that, so I'd probably prefer to let that handle it rather than duplicating the effort - it has more context to return a good error anyways, and is guaranteed to get called (unlike this function)

Reply:
+1 to Danny's comment. It's not necessary to duplicate the analysis. Users *could* do things that aren't useful to them, but in practice likely won't. 

The big risk with these is that users change types and such, and end up not registering new ones. But we can ultimately give them a clear non-failing message for all the types their pipeline uses that are missing with code to copy paste, for the perf boost.